### PR TITLE
Two factor authentication for node.js apps blog post.

### PR DESCRIPTION
**Pull request by sebadoom**
Ready for comments.

### CODE REVIEW

**Comment by sebadoom** [comment]
Full screen capture

**Comment by woloski** [comment]
Feedback
* We should have an aside at the end showing how it works on Auth0 (dashboard with a switch :))
* We should mention all the things that you have to implement for this to be production ready:
   - have a way to bypass this if you lost/change your phone (recovery codes)
   - have a way to reset the MFA for an admin user
   - fallback to SMS if the user does not have a smart phone
   - send emails every time mfa is turned on or off for your app
   - checkbox to remember the MFA for X days

**Comment by jgelsey** [comment]
Typo: s/enrolment/enrollment/g

File: _posts/2015-08-14-add-two-factor-authentication-to-your-node-apps-with-passport.markdown [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+```
+
+As usual, passport will call this callback with the actual 'user' object as deserialized from the request. In this case, we are using a simple JSON-based filesystem storage. In production, you should never store passwords and secrets as plain text. The 'user' object contains a 'key' string that may be null. In case the key is null, we will consider TOTP disabled and allow logins with just a username and password. In case the key is present, we will require an additional input from the user before proceeding with the login. The key must be provided to passport as a bytearray, but for convenience (and because Google Authenticator requires so) we store it as a BASE32 encoded string. The 'period' is the amount of time a single TOTP code es valid from the moment it is requested to the moment it is entered.
+
+The enrolment step requires the user to be already logged-in (as this is necessary to enable of disable TOTP in the server for each specific account). If the user chooses to enable TOTP, we present the shared-secret as a QR code ready to scan and save in Google Authenticator:
+
+```javascript
+app.get('/totp-setup', 
+    isLoggedIn,
+    ensureTotp,
```
Comment (mgonto) [inline _posts/2015-08-14-add-two-factor-authentication-to-your-node-apps-with-passport.markdown]: What's this `ensureTopt`?

Comment (mgonto) [inline _posts/2015-08-14-add-two-factor-authentication-to-your-node-apps-with-passport.markdown]: I'd explain it in words here (I see the code is below)

**Comment by mgonto** [comment]
I really like it :).

Agreed with @woloski. Add a section on "It's implemented, now what?" or something like that and mention all the steps to get it production ready. And then put an `Aside` that shows how easy it's to set it up with Auth0, with a link to the Multifactor section in the dashboard and a screenshot showing Mutifactor enabled

**Comment by mgonto** [comment]
@vctrfrnndz style for the post header and icon look good? Can you help with tha?

**Comment by mgonto** [comment]
@jgelsey after the blogpost is reviewed, we send it to wordy.com, so all typos will be fixed :).

**Comment by sebadoom** [comment]
@jgelsey Thanks for checking, I wasn't sure about it either. I did check the spelling: http://dictionary.reference.com/browse/enrolment, but I'll use whatever is commonplace :) I used one "l" because google seemed to like it more that way: https://www.google.com.ar/webhp?sourceid=chrome-instant&ion=1&espv=2&ie=UTF-8#q=define%20enrollment

**Comment by sebadoom** [comment]
Now awaiting more feedback :)

**Comment by sebadoom** [comment]
Can't post a full page screenshot because the extension is rendering the page really bad :( I'll see if another extension works better and post a capture if possible.

**Comment by sebadoom** [comment]
Pay no mind to the image in the aside section, it's a GIF, so it gets corrupted when captured. Sigh.
Full page capture

**Comment by mgonto** [comment]
All links to CDN should be `cdn.auth0.com` instead of `http://assets.auth0.com/`. They work!

**Comment by mgonto** [comment]
Other than that, I think it's good to go :)

**Comment by mgonto** [comment]
@sebadoom did you send it to wordy?

**Comment by sebadoom** [comment]
All changes done.
adding two-factor authentication to your node js express apps with passport js and google authenticator

**Comment by woloski** [comment]
screenshots @mgonto

**Comment by woloski** [comment]
too much text in some areas imo, can we try to cut some?

**Comment by mgonto** [comment]
For the MFA screenshot, use https://cdn.auth0.com/blog/grow-revenue/mfa.png.

**Comment by sebadoom** [comment]
@mgonto The GIF shows that part of the process. Just to confirm, do you still want to change it for the one you linked?

@woloski Let me see what I can do

**Comment by mgonto** [comment]
Yep, still to the image.

**Comment by woloski** [comment]
I would be more subtle. After the conclusion. "Aside: How it works in Auth0"

**Comment by woloski** [comment]
And use bullets to explain the value prop

* Turn on/off, no changes to your application
* Integrates with any TOTP app (Google Authenticator, Microsoft Authenticator, Authy, ....)
...

**Comment by dschenkelman** [comment]
Should we wait until I update the mfa widget with the new look?

**Comment by mgonto** [comment]
@dschenkelman I don't think so

**Comment by sebadoom** [comment]
We can update the picture later, I suppose.

**Comment by sebadoom** [comment]
Implemented suggestions.
adding two-factor authentication to your node js express apps with passport js and google authenticator

**Comment by sebadoom** [comment]
Err, sorry about that, missed the button. Sigh.

**Comment by woloski** [comment]
"First steps to production" should be "Is this production ready?"

**Comment by sebadoom** [comment]
from theory to practice adding two-factor to node js

**Comment by sebadoom** [comment]
Sent to wordy, awaiting results.

**Comment by woloski** [comment]
from a high level looks much better

**Comment by sebadoom** [comment]
Wordy changes implemented.
from theory to practice adding two-factor to node js 2

**Pull request merged**